### PR TITLE
[splash] handle splash as regular window

### PR DESCRIPTION
**Pull request by mkortstiege**
This fixes some flicker and render issues with pvr windows and the startup masterlock. Needs win32 sync and testing ;)

Issue @ http://forum.kodi.tv/showthread.php?tid=233147

### CODE REVIEW

**Comment by afedchin** [comment]
@mkortstiege here you go https://github.com/afedchin/xbmc/commit/3df8ca143cfbf6e73095a6a97640ce5074915ff9

**Comment by afedchin** [comment]
no splash at startup (tested on windows)

**Comment by mkortstiege** [comment]
@afedchin thanks. jenkins build this please.

**Comment by xhaggi** [comment]
tested on osx works fine for both (masterlock + pvr window as startup window)

**Comment by mkortstiege** [comment]
@afedchin that's odd but i guess related to the fact we're initing the windowing system too late. The old splash was triggered earlier. Not sure how to fix it with this approach.

**Comment by afedchin** [comment]
I think we should keep both splashes, one is image which is shown on startup and splash window which will replace splash image when windows manager is ready.

File: xbmc/Application.cpp [hunk 1]
```diff
@@ -1156,6 +1138,14 @@ bool CApplication::Initialize()
   g_curlInterface.Load();
   g_curlInterface.Unload();
 
+  if (g_windowManager.Initialized())
+  {
+    g_windowManager.CreateWindows();
+
+    if (g_advancedSettings.m_splashImage)
+      g_windowManager.ActivateWindow(WINDOW_SPLASH);
```
Comment (xhaggi) [inline xbmc/Application.cpp]: move this 2 lines directly before https://github.com/xbmc/xbmc/pull/7650/files#diff-63246197ac0f0ac89358c1155ad21d5bR822 

`g_windowManager.Initialize()` will call `LoadNotOnDemandWindows()` which then holds the splash window.

```
g_windowManager.Add(new CGUIWindowSplash);
g_windowManager.Initialize();
```

**Comment by afedchin** [comment]
something like this https://github.com/afedchin/xbmc/commit/0a153898f776ecff2958c4e24d51a17accbcd901

File: xbmc/Application.cpp (hunk 1 above)
Comment (xhaggi) [inline xbmc/Application.cpp]: and revert everything you've changed here except the `g_windowManager.Delete(WINDOW_SPLASH);`

File: xbmc/windows/GUIWindowSplash.cpp [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+#include "GUIWindowSplash.h"
+#include "filesystem/File.h"
+#include "filesystem/SpecialProtocol.h"
+#include "guilib/GUIImage.h"
+#include "guilib/GUIWindowManager.h"
+#include "utils/log.h"
+
+CGUIWindowSplash::CGUIWindowSplash(void) : CGUIWindow(WINDOW_SPLASH, "")
+{
+  m_image = nullptr;
```
Comment (xhaggi) [inline xbmc/windows/GUIWindowSplash.cpp]: change the `m_loadType` to be `LOAD_ON_GUI_INIT`

File: xbmc/Application.cpp (hunk 1 above)
Comment (xhaggi) [inline xbmc/Application.cpp]: BTW you could also check if we could init the window manager before loading the keymaps.

File: xbmc/windows/GUIWindowSplash.cpp (hunk 2 above)
Comment (xhaggi) [inline xbmc/windows/GUIWindowSplash.cpp]: without it'll not loaded by `CGUIWindowManager::LoadNotOnDemandWindows()`

File: xbmc/Application.cpp (hunk 1 above)
Comment (xhaggi) [inline xbmc/Application.cpp]: meh the link won't work. I meant line 822 in `Application.cpp`

File: xbmc/guilib/GUIWindowManager.cpp [hunk 3]
```diff
@@ -300,12 +301,14 @@ void CGUIWindowManager::CreateWindows()
   Add(new CGUIWindowScreensaver);
   Add(new CGUIWindowWeather);
   Add(new CGUIWindowStartup);
+  Add(new CGUIWindowSplash);
```
Comment (xhaggi) [inline xbmc/guilib/GUIWindowManager.cpp]: you can remove this line

**Comment by xhaggi** [comment]
@afedchin I've pushed a fix which should solve the rendering issue of the splash window. could you please test if it works. We are not really sure if the approach is the right one, but we could try ;)

**Comment by afedchin** [comment]
I'll try today.

**Comment by xhaggi** [comment]
we have thought about it again and now we leave the splash screen as it is, because the window needs some seconds to spawn instead of the splash screen. @afedchin now we have both as you suggested, the splash screen and the splash window which kicks in after the splash screen disappears.

**Comment by MartijnKaijser** [comment]
side question. Does now also remove the silly limitation of the fixed splash screensize?

**Comment by Memphiz** [comment]
Yes it does if i got it right. I also think that this would allow animated splash screens with current gif implementation ...

**Comment by afedchin** [comment]
@xhaggi It works. but there is a little difference between displaying the splash screen and the splash window. 
CSplash
CSplash
CGUIWindowSplash (I've changed position of dialog to out of screen to see difference)
CGUIWindowSplash

the splash window looks smaller than the splash screen

**Comment by xhaggi** [comment]
@afedchin yep because the resolution in CSplash is hard-coded to 720p. Let me check if it's possible to remove that.

**Comment by xhaggi** [comment]
@afedchin done in the last commit, please re-try it ;)

**Comment by afedchin** [comment]
In full screen now they look identical. In windowed mode there is a little in aspect ratio difference, but not critical

**Comment by xhaggi** [comment]
@afedchin another try please

**Comment by afedchin** [comment]
the latest commit broke splash image. now it's not displayed.

**Comment by xhaggi** [comment]
thanks, will revert it ;)

**Comment by xhaggi** [comment]
@afedchin i've fixed the scaling issue, please give it another try ;)

**Comment by xhaggi** [comment]
> side question. Does now also remove the silly limitation of the fixed splash screensize?

yep that's gone now.

File: xbmc/Application.cpp [hunk 4]
```diff
@@ -1171,6 +1158,12 @@ bool CApplication::Initialize()
     g_windowManager.CreateWindows();
     /* window id's 3000 - 3100 are reserved for python */
 
+    // initialize splash window after splash screen disappears
+    // because we need a full window in the background which gets
```
Comment (xhaggi) [inline xbmc/Application.cpp]: `real window` instead of `full window`

Comment (mkortstiege) [inline xbmc/Application.cpp]: you're commenting on your own changes now? ;)

Comment (xhaggi) [inline xbmc/Application.cpp]: As a reminder for my own mistakes ;)

**Comment by xhaggi** [comment]
squashed everything a bit together. @afedchin it would be nice if you could do another test round. the splash image will be scaled in order to get rid of the hard-coded size limitation. We need to update the splash to 1080p.

**Comment by afedchin** [comment]
@xhaggi with a couple of my fixes https://github.com/afedchin/xbmc/commits/splash_vs_sync it work perfectly :)

**Comment by mkortstiege** [comment]
@afedchin why the `AR_CENTER` revert? Reason we added was to get rid of the hard-coded values and have proper scaling.

**Comment by xhaggi** [comment]
@afedchin thx for the first commit which fix the resizing issue we've noticed ;)
The second one will be left-out because we want the image to fit the entire window.

**Comment by afedchin** [comment]
@mkortstiege in my oppinion with AR_CENTER we can use any splash which will no be scaled and always will be shown in center

**Comment by xhaggi** [comment]
@afedchin if you use a 1080p splash in window mode (lower 1080p) this will show you only a partial of the entire image.

**Comment by xhaggi** [comment]
jenkins build this please

**Comment by mkortstiege** [comment]
Updated the splash imge with a non resized version. Thanks @da-anda ;)

**Comment by da-anda** [comment]
it's 1920x1200 for now, but I also have a 4K one. Also I consider it temporary, as fullscreen splashes offer way much more possibilities that I need to experiment with. Thanks for finally allowing properly scaled fullscreen splashes :)

**Comment by xhaggi** [comment]
should we consider to do a backport because it fixes different issues?

**Comment by da-anda** [comment]
for a backport we'd have to support old splash size (non-stretched, centered) as we might encouter sizing issues or other oddities with fullscreen splashes, and I'd not risk that on 15.x where it's harder to fix than in nightlies

**Comment by xhaggi** [comment]
@da-anda we could use CENTER for the backport instead of SCALE.

**Comment by xhaggi** [comment]
apart from the backport discussion any objections or good to go?

**Comment by mkortstiege** [comment]
jenkins build this please

**Comment by koying** [comment]
+1 for droid

**Comment by MartijnKaijser** [comment]
jenkins build this please

**Comment by mkortstiege** [comment]
Win32 issue is unrelated (addon failed to build).

**Comment by MilhouseVH** [comment]
This PR may be causing odd behaviour with third-party skins (ie. not Confluence), or when splash is disabled in as.xml - see [posts from Mfleigle](http://forum.kodi.tv/showthread.php?tid=231092&pid=2069720#pid2069720) where the Home screen doesn't appear recruit when using Confluence, and when disabling splash "the OS freeze".

I'm on the road until late Monday afternoon and can investigate further once I'm home, unless anyone beats me to it.

**Comment by MilhouseVH** [comment]
I've knocked up a debug-enabled build (OpenELEC, RPi2) and have reproduced the crash with stock Confluence when splash is disabled in as.xml: http://sprunge.us/eJFP

**Comment by MilhouseVH** [comment]
Just tried an Ubuntu x64 Kodi build and this isn't crashing (though I'm running it remotely and I've no idea what is happening on screen), but maybe the crash is OpenELEC (or RPi, or even EGL) specific? If the OpenELEC line numbers from the crash log don't make sense I can provide the corrections tomorrow evening as the line numbers might have shifted around due to other patches present in my test build.

**Comment by mkortstiege** [comment]
@MilhouseVH, please give https://gist.github.com/mkortstiege/5a1c8139dc2ed9193ae6 a try.

**Pull request merged**